### PR TITLE
fix streaming interface for S3

### PR DESCRIPTION
**Pull request by laborg**
I had problems when trying to stream the data of large files in S3 into local files and discovered a bug as well as an suboptimal implementation.

- `response_stream` expects a `Base.BufferStream` but within `get_objects` this is then turned into a query part, and of course this doesn't work well.
- Most of the Julia IO stuff excepts <:IO because this can easily be provided with `open()`. Having a more general type above `Base.BufferStream` allows for such flexibility.

### CODE REVIEW

File: src/AWS.jl [hunk 1]
```diff
@@ -575,8 +575,10 @@ function (service::RestXMLService)(
         request.response_dict_type = pop!(args, "response_dict_type")
     end
 
-    delete!(args, "headers")
-    delete!(args, "body")
+    for key in ("headers", "body", "response_stream", "return_stream", "return_raw")
+        delete!(args, key)
+    end
```
Comment (mattBrzezinski) [inline src/AWS.jl:580]: Looking at this some more, we can do one even better. When making the Request object above and setting these parameters we can just do this for each of the keys:

> pop!(args, "headers", [])

File: test/minio.jl [hunk 2]
```diff
@@ -32,6 +32,17 @@ S3.put_object("anewbucket","foo/baz",Dict("body"=>"a secondnested object"))
...
+# Fix: BufferStream can be generalized to IO
+# v1.29.1: ERROR: MethodError: Cannot `convert` an object of type IOStream to an object of type Base.BufferStream
+mktemp() do f, io
+    S3.get_object("anewbucket","myobject", Dict("response_stream"=>io))
+    @test read(f, String) == "Hi from Minio"
+end
+
+# Bug: `response_stream` is pushed into the query part...
+# v1.29.1: ERROR: MethodError: no method matching iterate(::Base.BufferStream)
+S3.get_object("anewbucket","myobject", Dict("response_stream"=>Base.BufferStream(), "return_stream"=>true))
```
Comment (mattBrzezinski) [inline test/minio.jl:44]:
```suggestion
S3.get_object("anewbucket", "myobject", Dict("response_stream"=>Base.BufferStream(), "return_stream"=>true))
```

File: test/minio.jl [hunk 3]
```diff
@@ -32,6 +32,17 @@ S3.put_object("anewbucket","foo/baz",Dict("body"=>"a secondnested object"))
 # Test retrieving an object
 @test S3.get_object("anewbucket","myobject") |> String == "Hi from Minio"
 
+# Fix: BufferStream can be generalized to IO
```
Comment (mattBrzezinski) [inline test/minio.jl:35]: These can probably get moved into the `test/AWS.jl` file as they're general tests and not Min.IO specific. Although with that said some general tests need to be in place for Min.IO to check it's functionality works too.

File: src/AWS.jl (hunk 1 above)
Comment (laborg) [reply src/AWS.jl]: Yes, that looks better.

File: test/minio.jl (hunk 3 above)
Comment (laborg) [reply test/minio.jl]: I've reshuffled this a bit: 
 - A "streaming" test is now in AWS.jl and minio.jl, as both should handle this case properly.
 - A regression test in issues.jl concerning the problem where the `response_stream` was pushed in to the query.

File: test/issues.jl [hunk 4]
```diff
@@ -49,3 +49,24 @@ end
         @test_throws NoCredentials Lambda.list_functions(;aws_config=AWSConfig(creds=nothing))
     end
 end
+
+@testset "issue streaming" begin
```
Comment (mattBrzezinski) [inline test/issues.jl:53]: Would you mind cutting an issue explaining the problem then titling this test set `issue XYZ`?

File: src/AWS.jl [hunk 5]
```diff
@@ -563,20 +563,15 @@ function (service::RestXMLService)(
         service=service.name,
         api_version=service.api_version,
         request_method=request_method,
-        content=get(args, "body", ""),
-        headers=LittleDict{String, String}(get(args, "headers", [])),
-        return_stream=get(args, "return_stream", false),
-        http_options=get(args, "http_options", []),
-        response_stream=get(args, "response_stream", nothing),
-        return_raw=get(args, "return_raw", false),
+        content=pop!(args, "body", ""),
```
Comment (mattBrzezinski) [inline src/AWS.jl:566]: A bit out of scope, but do you mind changing the RESTJSON, JSON and Query functors as well to this `pop!` style? I think they should be consistent throughout.

Comment (laborg) [reply src/AWS.jl:570]: I've refactored the code a little to get rid of some duplications.

File: src/AWS.jl [hunk 6]
```diff
@@ -465,15 +477,7 @@ function _generate_rest_resource(request_uri::String, args::AbstractDict{String,
...
-function _return_headers(args::AbstractDict{String, <:Any})
-    return_headers = get(args, "return_headers", false)
-
-    if return_headers
-        delete!(args, "return_headers")
-    end
-
-    return return_headers
-end
+_return_headers(args::AbstractDict{String, <:Any}) = pop!(args, "return_headers", false)
```
Comment (mattBrzezinski) [inline src/AWS.jl:480]: We can just scrap this function and call `pop!` where applicable

File: src/AWS.jl (hunk 5 above)
Comment (mattBrzezinski) [reply src/AWS.jl:570]: Awesome thank you!

**Review by mattBrzezinski — COMMENTED** [review]
bors try

File: test/issues.jl [hunk 7]
```diff
@@ -49,3 +49,24 @@ end
         @test_throws NoCredentials Lambda.list_functions(;aws_config=AWSConfig(creds=nothing))
     end
 end
+
+@testset "issue 324" begin
+    body = "Hello World!"
+    file_name = "streaming.bin"
+
+    try
+        S3.put_object(bucket_name, file_name, Dict("body" => body))
```
Comment (mattBrzezinski) [inline test/issues.jl:58]: Ah looks like when I made this I created and deleted the bucket after every test.
Probably to ensure that no artifacts get stuck and left over.

You'll just need to preface this line with:

```suggestion
        S3.create_bucket(bucket_name)
        S3.put_object(bucket_name, file_name, Dict("body" => body))
```

Comment (laborg) [reply test/issues.jl:59]: done

**Comment by mattBrzezinski** [comment]
bors try

File: src/AWS.jl (hunk 6 above)
Comment (laborg) [reply src/AWS.jl]: I removed the `_return_headers` test, because it was basically just testing the three arg `pop!()` call method.

**Comment by mattBrzezinski** [comment]
Ugh... LittleDicts do not implement `pop!` with a default value. I'll make these changes in `OrderedCollections`, feel free to revert that change and I can get around to putting it back in.

**Comment by mattBrzezinski** [comment]
Oh it already looks like I was trying to do this in the past...

> https://github.com/JuliaCollections/OrderedCollections.jl/pull/59

**Comment by laborg** [comment]
Do you see a problem with the proposed `_pop!` as a fall back for now?

File: src/AWS.jl [hunk 8]
```diff
@@ -722,32 +703,21 @@ function (service::RestJSONService)(
...
     aws_config::AbstractAWSConfig=global_aws_config(),
 )
-    return_headers = _return_headers(args)
+    return_headers = _pop!(args, "return_headers", false)
 
-    request = Request(
-        service=service.name,
-        api_version=service.api_version,
+    request = Request(;
+        _extract_common_kw_args(service,args)...,
```
Comment (mattBrzezinski) [inline src/AWS.jl:709]:
```suggestion
        _extract_common_kw_args(service, args)...,
```

File: src/AWS.jl [hunk 9]
```diff
@@ -162,12 +162,36 @@ Base.@kwdef mutable struct Request
...
     response_dict_type::Type{<:AbstractDict}=LittleDict
 end
 
+# use this until the threed arg pop! is available for LittleDict
+# https://github.com/JuliaCollections/OrderedCollections.jl/pull/59
+function _pop!(dict::AbstractDict{String, <: Any},kw,default)
+    if haskey(dict,kw)
+        val = dict[kw]
+        delete!(dict,kw)
+        return val
```
Comment (mattBrzezinski) [inline src/AWS.jl:177]:
```suggestion

        return val
```

**Comment by mattBrzezinski** [comment]
bors try

**Review by mattBrzezinski — COMMENTED** [review]
Two minor white spacing things.

**Comment by mattBrzezinski** [comment]
> Do you see a problem with the proposed `_pop!` as a fall back for now?

No it makes sense, I don't see this being problematic.

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try

**Comment by laborg** [comment]
Anything else left to do?

**Review by mattBrzezinski — APPROVED** [review]
bors r+

**Pull request merged**